### DISCUSSION TITLE
Is it possible to utilize `freeze` in a custom Lua filter? (for the sake of rendering TikZ figures)

### DISCUSSION THREAD

**Discussion by sun123zxy** [quarto-dev/quarto-cli#11517] [Q&A]
### Description

I'm trying to implement a custom Lua filter to render TikZ code of the form
`````
```{tikz}
\begin{tikzfigure}
% anything
\end{tikzfigure}
```
`````
into an embedded figure in HTML, just like what a regular computational figure works. It's not too hard to implement though, with the aid of [an example of Pandoc filter](https://pandoc.org/lua-filters.html#building-images-with-tikz) and dvisvgm. The major problem is that I failed to figure out how to `freeze` the computed result in a custom Lua filter, i.e. to preserve the generated `.svg` file and prevent re-rendering. Seemingly little documentation is provided on this topic. Any suggestions would be appreciated!

**Comment by mcanouil** [comment]
I don't think the [freeze](https://quarto.org/docs/projects/code-execution.html#freeze) logic is accessible outside of Quarto CLI currently (thus the fact that you don't find documentation).

FYI:

- https://github.com/pandoc-ext/diagram

**Comment by cderv** [comment]
There is also this extension about tikz: https://github.com/danmackinlay/quarto_tikz 

maybe it does already what you want ?

**Comment by sun123zxy** [comment]
Thanks for the recommendations above. Actually I tried diagram and quarto_tikz a few months ago. Sadly they were either outdated or crashed on my device at that time (not sure if they work now). Neither do they have a `freeze` logic, though. Currently I'm using my DIY filter without `freeze` and it works well.

**Reply by cderv** [reply]
`freeze` is a feature that targets computation. As of Quarto 1.6, we don't have a feature to extent computation engine. So even if you manage to do a Lua filter that does handle `tikz` as if it was a code cell, it will be a Lua filter logic when doing Pandoc processing, and not a computation engine logic as for R, python, julia or even diagrams. Those needs to be in Typescript codebase, and there is no way to have extension contribute an engine. 

To be clear, `freeze` needs to be handled at the engine computation stages, and before rendering with Pandoc using the Lua filter chain. 

There are several issues that are related to handling some cells in Lua though
- https://github.com/quarto-dev/quarto-cli/issues/10530 
- https://github.com/quarto-dev/quarto-cli/issues/4839

**Comment by jdutant** [comment]
[Imagify](https://github.com/dialoa/imagify ) might work for you. It's a Quarto / Pandoc filter that compiles tikz into svg. If you set it to save the SVG in separate files (perhaps in a folder that Quarto doesn't delete) it should only recreate them when the code has changed. It's not a sophisticated freeze method: it merely uses hashes of the tikz code as filenames, and since the hash is the same if the code stays the same, it simply avoids recreating files that already exist. 

https://github.com/dialoa/imagify
